server/etcdserver/v3_server.go

```diff
@@ -156,8 +168,30 @@ func (s *EtcdServer) DeleteRange(ctx context.Context, r *pb.DeleteRangeRequest)
 	return resp.(*pb.DeleteRangeResponse), nil
 }
 
+// firstCompareKey returns first non-empty key in the list of comparison operations.
+func firstCompareKey(c []*pb.Compare) string {
+	for _, op := range c {
+		key := string(op.GetKey())
+		if key != "" {
+			return key
+		}
+	}
+	return ""
+}
+
 func (s *EtcdServer) Txn(ctx context.Context, r *pb.TxnRequest) (*pb.TxnResponse, error) {
 	readOnly := txn.IsTxnReadonly(r)
+
+	var span trace.Span
+	ctx, span = traceutil.Tracer.Start(ctx, "txn", trace.WithAttributes(
+		attribute.String("first_compare_key", firstCompareKey(r.GetCompare())),
+		attribute.Int("compare_len", len(r.GetCompare())),
```

Comment:
Maybe add a type of comparison and value we compare?
Like `attribute.Int("compare_first_op", "equal")`, `attribute.Int("compare_first_source", "mod_rev")`, `attribute.Int("compare_first_value", "42")`,

Reply:
There are two problems:
1. `Result` defaults to `equal`, `Target` defaults to `version` making it slightly harder to implement as we need to ensure attributes are not added when there is no `compare` (is it required?)
2. `TargetUnion` is `oneof` and with the currently generated code, there is no nice way to get a string out of it; again, not impossible to overcome, but it could be better suited for another PR

Reply:
Ok, we can move it to followup.